britnay268/OutTheGC
# Stretch Part 1

## Description
Updated the database to handle user votes functionality which is a many to many between users and activities as well as added to attributes to the activity entity: isArchived and isDeleted. Further, created apis to cast a vote, archive an activity, mark an activity as deleted and get all archived and deleted  activities for a trip. Lastly, updated api response for trip details to return vote count as well as any other apis that returns activities. Lastly, adjusted endpoints to produce the correct data entity respective to its endpoint.

## Related Issue
- #26 
- #28 
- #29 
- #38

## Motivation and Context
To enhance the application's functionality and data consistency, the database was updated to support user votes through a many-to-many relationship between users and activities. Two attributes, isArchived and isDeleted, were added to the activity entity to better manage activity states. New APIs were implemented to enable voting, archiving, deleting, and retrieving archived or deleted activities for a trip. Additionally, the trip details API was improved to include vote counts, while other activity-related APIs were updated to provide accurate and consistent data. Endpoints were adjusted to ensure they return the correct data entities, aligning the responses with their intended functionality.

## How Can This Be Tested?
1. Pull down repository
2. Run the below code to create connection to the database
```
dotnet user-secrets init
dotnet user-secrets set "OutTheGCDbConnectionString" "Host=localhost;Port=5432;Username=postgres;Password=<your_postgresql_password>;Database=OutTheGC"
```
3. Run `dotnet ef database update`
4. Run the application and test all the following endpoints for improvements
     - GET activity/{activityId} to see activity vote count
     - PUT /activity/{activityId}/vote/{userId} to add or remove a vote from an activity
     - GET /activity/filter to filter activity by highest votes first
     - PATCH /activity/{activityId}/archive to archive an activity
     - PATCH /activity/{activityId}/delete to mark activity as deleted
     - GET /trip/{tripId}/activities/inactive to see all the activities on a trip that are archived or deleted if any
     - GET /trip/{tripId} to see all trip activities' vote counts 

## Types of changes
